### PR TITLE
Use `MutexLock` in more places

### PR DESCRIPTION
**Pull request by AThousandShips**
This ensures more safe management of mutexes and makes forgetting to unlock it less likely, most of these cases are trivial though and don't depend on it for safety but I'd prefer to avoid using direct locking unless it's necessary to handle locking between methods etc. or in some cases where using it is just too cluttered

There are some cases which could benefit from a little more of this using the new `temp_unlock/relock` methods of `MutexLock` but kept things simple here

### CODE REVIEW

File: core/io/resource.cpp [hunk 1]
```diff
@@ -60,32 +60,32 @@ void Resource::set_path(const String &p_path, bool p_take_over) {
 		p_take_over = false; // Can't take over an empty path
 	}
 
-	ResourceCache::lock.lock();
+	{
+		MutexLock lock(ResourceCache::lock);
```
Comment (AThousandShips) [inline core/io/resource.cpp:64]: We should probably rename `ResourceCache::lock` to `ResourceCache::mutex` at some point as it's just a recipe for confusion having an incorrect name for it, left it alone in this PR though

File: modules/navigation/2d/nav_mesh_generator_2d.cpp [hunk 2]
```diff
@@ -87,8 +87,8 @@ void NavMeshGenerator2D::sync() {
 		return;
 	}
 
-	baking_navmesh_mutex.lock();
-	generator_task_mutex.lock();
+	MutexLock baking_navmesh_lock(baking_navmesh_mutex);
+	MutexLock generator_task_lock(generator_task_mutex);
```
Comment (AThousandShips) [inline modules/navigation/2d/nav_mesh_generator_2d.cpp:91]: While it doesn't seem to cause any problems I feel that locking two mutexes like this is a recipe for starvation, so would be good to look into that down the line

**Comment by RandomShaper** [comment]
I'm a fan of consistency as much as I'm a fan of expressivity. So, on the one hand, I feel that we lose something by letting explicit lock-unlock go (because in some places, having the unlock line around helps remembering you're reading a critical section, etc.)., while on the other hand, I feel that trying to use the RAII syntax always reduces cognitive load when maintaining code as you don't need to ask yourself which one to use.

I'd say the tradeoff is a bit towards the bad side, but that's a bit subjective, so let's hear more opinions.

**Comment by AThousandShips** [comment]
Can split up the cases that are trailing etc. and keeping it in trivial cases like a full block

**Comment by RandomShaper** [comment]
> Can split up the cases that are trailing etc. and keeping it in trivial cases like a full block especially with newly indicated blocks

That sounds very reasonable.

File: modules/navigation/2d/nav_mesh_generator_2d.cpp [hunk 3]
```diff
@@ -117,8 +117,8 @@ void NavMeshGenerator2D::sync() {
 }
 
 void NavMeshGenerator2D::cleanup() {
-	baking_navmesh_mutex.lock();
-	generator_task_mutex.lock();
+	MutexLock baking_navmesh_lock(baking_navmesh_mutex);
+	MutexLock generator_task_lock(generator_task_mutex);
```
Comment (RandomShaper) [inline modules/navigation/2d/nav_mesh_generator_2d.cpp:118]: Despite these will be unlocked in reverse order, matching the former code, I'd still put the second lock in a scope, for clarity.

File: core/io/resource.cpp [hunk 4]
```diff
@@ -600,7 +600,7 @@ void ResourceCache::clear() {
 }
 
 bool ResourceCache::has(const String &p_path) {
-	lock.lock();
+	MutexLock mutex_lock(lock);
```
Comment (RandomShaper) [inline core/io/resource.cpp:600]: I'd suggest putting this one in a scope in a way that it's unlocked before the `!res! check, to better follow the (implicit) rule of unlocking as soon as possible. In this case, a simple null check won't keep the mutex locked for a lot longer, though, but for consistency and micro-pessimization avoidance, I'd go for that.

**Review by RandomShaper — APPROVED** [review]
Looks great now! I added some suggestions, but I'm leaving an approval already.

Aside, if the commit structure is not meant to be squashed, I'd also suggest rewording the messages so they look very similar and it's obvious they part of the same endeavor.

**Comment by akien-mga** [comment]
Thanks!

**Comment by AThousandShips** [comment]
Thank *you*!
